### DISCUSSION TITLE
Prog Change and set Midi Channel

### DISCUSSION THREAD

**Discussion by Lenvdb64** [melanchall/drywetmidi#310] [Q&A] [answered]
Hi all
I am using the Patternbuilder to create a Drum pattern.
All the work will be done using the **General Midi Drums** on **Midi Channel 10**.
As it is 0-based I will need to select Channel 10 as  **Channel = 9**:
But I cannot see how to do this in PatternBuilder

My Code so far:
```
` private static Pattern CreatePattern() 
 {
     try
     {
         var pB = new PatternBuilder();            
       
 //How to set Midi Ch 10 ??
 pB.ControlChange((SevenBitNumber)1, (SevenBitNumber)1);
         
//set PC to select Prog 0 
pB.ProgramChange((SevenBitNumber)0);`
```

I am sure many of you have done this already.
Please could you kindly share how you did this?

**Comment by Lenvdb64** [comment]
Actually - no longer an issue

Here is my code I used:

 ```
using (var dev = OutputDevice.GetByName("Microsoft GS Wavetable Synth"))
 {
     Melanchall.DryWetMidi.Common.SevenBitNumber channel = (Melanchall.DryWetMidi.Common.SevenBitNumber)9;

     CreateMidiFile();
 }
```

**Comment by melanchall** [comment, marked as answer]
@Lenvdb64 Hi,

`Pattern` is just a set of actions. Those actions will be materialized on [`ToFile`](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Composing.Pattern.html#Melanchall_DryWetMidi_Composing_Pattern_ToFile_Melanchall_DryWetMidi_Interaction_TempoMap_Melanchall_DryWetMidi_Common_FourBitNumber_) or [`TrackChunk`](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Composing.Pattern.html#Melanchall_DryWetMidi_Composing_Pattern_ToTrackChunk_Melanchall_DryWetMidi_Interaction_TempoMap_Melanchall_DryWetMidi_Common_FourBitNumber_) call. Please see the [Pattern](https://melanchall.github.io/drywetmidi/articles/composing/Pattern.html) article:

> Pattern can be then saved to [MidiFile](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Core.MidiFile.html) (via [ToFile](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Composing.Pattern.html#Melanchall_DryWetMidi_Composing_Pattern_ToFile_) method) or [TrackChunk](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Core.TrackChunk.html) (via [ToTrackChunk](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Composing.Pattern.html#Melanchall_DryWetMidi_Composing_Pattern_ToTrackChunk_) method). You need to provide a [tempo map](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Interaction.TempoMap.html). Also you can optionally specify the channel that should be set to events. The default channel is 0.

So you specify a channel by passing it as the parameter to those methods.

As for your code:

> ```csharp
> using (var dev = OutputDevice.GetByName("Microsoft GS Wavetable Synth"))
> {
>     Melanchall.DryWetMidi.Common.SevenBitNumber channel = (Melanchall.DryWetMidi.Common.SevenBitNumber)9;
> 
>     CreateMidiFile();
> }
> ```

it's super strange:

1. `channel` variable is not used;
2. `dev` is not used;
3. you work with the device in `using` block which is can cause problems (after the program leaves the block, the device will not exist and your attempts to interact with it can broke the program).

Honestly I don't understand what you want to say with this code :-) But I'm glad you've solved the problem.

**Reply by Lenvdb64** [reply]
Thank you Maxim
Yes - CreateMidifile does a few of these steps and calls CreatePattern

 ```
 public static void CreateMidiFile()
  {
      // Create an empty MIDI file
      var tempoMap = TempoMap.Create(Tempo.FromBeatsPerMinute(giBPM));
      var midiFile = new MidiFile();
      midiFile.ReplaceTempoMap(tempoMap);
      //TODO: ProgramChangeEvent
      //TODO: Set to Channel 10 (9)
      //TODO: Set to GM Drums Program 0

      var pBMain = CreatePattern();
      midiFile = pBMain.ToFile(tempoMap); 
      midiFile.Write("C:\\Users\\lenvd\\Documents\\VSProjects\\MyMidiPanel2\\Ptn1_4_4_8.mid");
  }
```

Then CreatePattern does the actual Pattern in a PatternBuilder...

So From a Logical view all happens within this Codeblock.
But yes - I see that the Channel is not passed or set as I still only hear the Grand Piano from Channel 0 on the GS synth.

Moooo...

**Reply by melanchall** [reply]
@Lenvdb64 From my answer you know how to set a channel now :-)

**Comment by CIRCE-EYES** [comment]
Hello Len! Greetings from Spain.
 I see that you are making a kind of beatbox drums, using Maxim Drywetmidi's wonderful library. 
A while ago I did something similar using a datagrid as cells, it looked cool, but I'm looking for something more similar to a piano roll, I did it in vb.net. Maybe my example and other apps that I have made to learn the library can help you drywetmidi, Maxim is a very nice guy and helped me to a certain extent, first he doesn't know vb.net, and second he can't answer as many questions as I had. 
I would really like to see your progress on the project, and if you share it, to be able to use it.
 I repeat, I'm looking for someone who dares to do a simple piano roll project. Greetings !!!!!

Have a look here :   https://github.com/CIRCE-EYES/Roll-midi-notes-using-a-Datagridview-object-Vstudio

**Reply by Lenvdb64** [reply]
Hello! 
I have decided to use C# and have a Panel and use the Paint Event to draw a Midi Grid step editor, having rows and columns.
Behind the Panel I have a List Object that holds Boolean values for when the grid cell is selected or not.
It sounds rather elaborate but I got the idea from "Honey the Code Witch" who has a similar project but used buttons instead of a Panel, but also with a List object holding Boolean values for the cells selected.

![image](https://github.com/user-attachments/assets/4103cb80-be60-4770-8416-29021a2d5b91)

I obviously have a few issues as I have only used this Lib for a week now....
1) I need to select a Channel 10 for General Midi to access the GM Drumkit.
2) I loop through the underlying Boolean List to collect all the Notes selected and now struggle to translate these into Note Events on the DryWet Library and then write it as a Midi file.

![image](https://github.com/user-attachments/assets/1b869e9b-b9a2-4934-aee8-c292748100bc)


Creating the Midi file is going well, in principle, and later I will add a proper FileSave dialog for it.
Notes and being generated but as stated - It is not selecting the GM Drum kit on Channel 10.
I am currently using a PatternBuilder and I build a Pattern, which can be saved to a Midi File.
When I play it back it is just Piano Notes.
As a standard Cubase has set the PPQN (Pixels per Quarter Note - to 960
and Ticks per Beat is also 960
Most Libs do not quite use this as a standard.

This is what I have so far.
I am hoping to use the C# VST2.Net lib to publish this as a VST to use with Monster Drum Kits....
I will take a look at your VB Code - Thank you for sharing!

**Reply by melanchall** [reply]
@Lenvdb64 You wrote:

> Notes and being generated but as stated - It is not selecting the GM Drum kit on Channel 10.

Did you read my first answer? I've already explained how to set a channel. You don't know how to pass a parameter to `ToFile` method?

```csharp
var pBMain = CreatePattern();
midiFile = pBMain.ToFile(tempoMap, (FourBitNumber)9); 
```
